File: algorithm/530. Minimum Absolute Difference in BST/530.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.val,self.diff=-1,2147483647
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root.left:
            self.getMinimumDifference(root.left)
        if(self.val>=0):
            self.diff=min(self.diff,root.val-self.val)
        self.val=root.val
        if root.right:
            self.getMinimumDifference(root.right)
        return self.diff

File: algorithm/530. Minimum Absolute Difference in BST/530.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    TreeNode prev;
    int diff = Integer.MAX_VALUE;
    public int getMinimumDifference(TreeNode root) {
        helper(root);
        return diff;
    }
    
    private void helper(TreeNode root) {
        if(root == null)
            return;
        helper(root.left);
        if(prev != null)    diff = Math.min(diff, root.val - prev.val);
        prev = root;
        helper(root.right);
    }
    
}
